### DISCUSSION TITLE
ESPNow Acknowledge; broadcast vs peer-to-peer

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11705] [General]
Currently using ESPnow in peer-to-peer, where a sender receives a message acknowledge from the receiver.

If a sender sends a broadcast to many receivers, do all receivers send acknowledges, or receivers do not acknowledge broadcast messages?

UPDATE: Seems I've found an answer to my question; broadcast messages are not acknowledged (which makes sense, as ACK transmissions from multiple receivers would collide):
https://www.instructables.com/ESPNow-a-Deeper-Look-Unicast-Vs-Broadcast-ACKs-Ret/

**Comment by glenn20** [comment]
Yes - as you have identified, broadcast messages are not acknowledged by the receiver. The `send()` always returns `True`. 

As a consequence, broadcast messages are never re-transmitted. Unicast message will be retransmitted many times if no ack is received. This means that broadcast messages can be *much* less reliable in noisy environments.

**Reply by glenn20** [reply]
Oh - and just to clarify, besides sending to the broadcast (or a multicast) address, you can `e.send(None, msg)` which will send the message to all the peers registered (with `add_peer()`). These will be sent as unicast messages with acknowledgement from each peer and re-transmissions if necessary. ie. it acts like a broadcast to your group of peers, but more reliably and helps keep the local wifi broadcast traffic down.

**Reply by mmliam** [reply]
`e.send(None, msg)` That's great to know, thanks glenn.

BTW: For ESP8266 modules (specifically ESP-07),  no automatic retransmissions are sent on no ACK. When the e.send(unit, msg) returns false my code waits one second then re-transmits (up to 2 more times). If the ESP32 receives any retransmissions, they are separated by one second.  Only a maximum of 3 transmissions are ever received by the ESP32 for a single event.
It's probably a characteristic of the ESP32 ESPNow code to do automatic retransmissions, on no ACK.

**Reply by glenn20** [reply]
I didn't realise that about the esp8266. The retransmission behaviour is poorly documented and comes from reports from people who have sniffed the wifi packets.

You can clearly see the retransmissions for the esp32 on the 3rd graph down in https://github.com/glenn20/upy-esp32-experiments/blob/main/ESPNowvsWifiEnergyUsage/README.md. The retransmissions are automatically sent within a millisecond or so if no ack is received. 

I haven't done the same tests with the esp8266. How do you know the esp8266 does not send retransmissions on no ack? I haven't found it in any documentation so far.

**Reply by glenn20** [reply]
In fact, this video would suggest that retransmissions do occur with esp8266 as well: https://www.youtube.com/watch?v=Xmzk6v5qIjo.

**Reply by mmliam** [reply]
All I can say is the ESP8266 sends to an ESP32. I can monitor what the ESP32 receives in an IDE as I have it printout all received messages. The ESP32 receives all the transmissions from the ESP8266, but sometimes the ESP8266 does not get the ACK and transmits again (up to 2 more times). All retransmissions are separated by one second, as per my coding. If the ESP8266 were sending out retransmission bursts I would see that output to the IDE from the ESP32. The most transmissions the ESP32 receives from the the ESP8266 is the maximum of 3 per trigger event.

Also, in monitoring the ESP8266 sends I've found that the send return is a fixed 25ms, whether it returns true (ACK), or false (NACK). This would be consistent with ESPNow maintaining a deterministic synchronization between all senders.

Perhaps it's the version of micropython [v1.191(583K)] that I'm using. Frankly, I'm glad I can control the retransmissions, so I don't waste battery power (3 transmissions MAX). I wouldn't like it if ESPNow were automatically re-transmitting.

I noted in the video he was supposedly seeing ACK's during broadcast that he couldn't explain; I'm beginning to suspect WireShark. Also, the code he's using is Arduino (a C++ derivative), perhaps the ESPNow implementation is different than the micropython implementation for the ESP8266. 

I was looking over the Espressif documentation for ESPNow, and so far have not found any discussion on automatic retransmissions. However, it does talk about the user coding a re-transmit procedure and assigning a sequence number in the message, so the receiver can filter out extra transmissions. This certainly doesn't sound like the ESPNow protocol does automatic retransmissions. Perhaps some implementation in a given language might include automatic retransmissions, but that would not be part of (or in my case desirable) ESPNow.

From the Espressif ESPNow documentation:

>  If receiving ack data timeouts, retransmit the ESP-NOW data. A sequence number can also be assigned to ESP-NOW data to drop the duplicate data.

BTW: I do sequence my transmissions, so I know exactly which transmissions from the ESP8266 were received by the ESP32 for a given event. I never get multiple receipts of a message with the same sequence number, as would be the case if the send routine was automatically re-transmitting.

**Reply by glenn20** [reply]
> All retransmissions are separated by one second, as per my coding. If the ESP8266 were sending out retransmission bursts I would see that output to the IDE from the ESP32.

Ah, now I underdstand. The espnow retransmissions I am talking about are automatically performed deep down in the Espressif software stack and are a feature of the ESPNOW protocol. You will not see evidence of them in the IDE. So, whenever you get a failed `send()` it has actually tried to re-send the message multiple times within 25 milliseconds before `send()` returns.

> Also, in monitoring the ESP8266 sends I've found that the send return is a fixed 25ms, whether it returns true (ACK), or false (NACK). This would be consistent with ESPNow maintaining a deterministic synchronization between all senders.

Have you tried this with a recent version of the espnow images? As I have mentioned previously, older versions were implemented so that there was an enforced busy-wait for at least 25ms. Newer versions return within a few ms if the send was successful on the first transmit.

> Perhaps it's the version of micropython [v1.191(583K)] that I'm using. Frankly, I'm glad I can control the retransmissions, so I don't waste battery power (3 transmissions MAX). I wouldn't like it if ESPNow were automatically re-transmitting.

Sorry to tell you that I believe it is (confirmed on ESP32 and others have reported it for the ESP8266) - and we have no control over that. It is inbuilt to the ESPNow protocol.

Note that these retransmissions are completely transparent to the user and I don't believe it makes any difference to your usage.

> I noted in the video he was supposedly seeing ACK's during broadcast that he couldn't explain; I'm beginning to suspect WireShark. Also, the code he's using is Arduino (a C++ derivative), perhaps the ESPNow implementation is different than the micropython implementation for the ESP8266.

See also https://hackaday.io/project/161896-linux-espnow/log/154698-how-esp-now-really-works for additional evidence for the esp8266.

> I was looking over the Espressif documentation for ESPNow, and so far have not found any discussion on automatic retransmissions. However, it does talk about the user coding a re-transmit procedure and assigning a sequence number in the message, so the receiver can filter out extra transmissions. 

This is to address the fact that despite the fact an ACK has been received from the recipient, the packet MAY have been dropped in the application layer (eg. on micropython if the internal rxbuffer buffer is full).

> BTW: I do sequence my transmissions, so I know exactly which transmissions from the ESP8266 were received by the ESP32 for a given event. I never get multiple receipts of a message with the same sequence number, as would be the case if the send routine was automatically re-transmitting.

**Reply by glenn20** [reply]
Oh - and...

> Perhaps it's the version of micropython [v1.191(583K)] that I'm using. Frankly, I'm glad I can control the retransmissions, so I don't waste battery power (3 transmissions MAX).

https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowvsWifiEnergyUsage shows that a failed `send()` consumes an extra 11 microWatt-hours of energy over a successful `send()`.

**Reply by mmliam** [reply]
Well now I understand why I'm not seeing the additional transmissions.

> See also https://hackaday.io/project/161896-linux-espnow/log/154698-how-esp-now-really-works for additional evidence for the esp8266.

From the above:

> First bad news: ESP-NOW continues to send the packet again and again for 20ms !
> 
> I hope their is a way to disable this resend behavior... Indeed, for a robotic control, we prefer to wait for the next up-to-date packet than getting an outdated one, and jamming the channel
> 

I agree; I hope there is away to disable the re-sends.

> In fact, this video would suggest that retransmissions do occur with esp8266 as well: https://www.youtube.com/watch?v=Xmzk6v5qIjo.

From the video:
11 re-transmissions per send x 11 uW = (0.121mw per send) x (how many sends per day) [in my case 50] = 6mw x 365 days = 2.2watts. How long is that battery going to last? For the battery I'm using only 10 months, was hoping for 2.5 years. Without the burst extraneous re-transmissions I'd get way more than 2.5 years.

For my application where the sender is located there is wifi noise that causes the sender to fail getting ACK's so it does a lot of my programmed re-sends.

However, this has been a most enlightening discussion.  I'm still amazed I can find nothing from the Esspressif ESPNow documentation that speaks to this important issue of automatic re-transmissions. Perhaps others have noticed unexpected battery drain, so the turned to Wire-Shark.  Looks like I need to update and start using my Wire-Shark - lol

**Comment by glenn20** [comment]
>
> I agree; I hope there is away to disable the re-sends.
>
> There is not.

> From the video:
> 11 re-transmissions per send x 11 uW
>
It is 11uWh for all the 11 retransmissions and they are only sent if no ack
to first message.

**Reply by mmliam** [reply]
> It is 11uWh for all the 11 retransmissions and they are only sent if no ack
> to first message.
> 
I was hoping you were going to say that.

Regarding https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowvsWifiEnergyUsage
There's something puzzling about this table entry, specifically the "Send" times:

<html><body>


Test | Time (ms) | Charge (mC) | Energy (microWh)
-- | -- | -- | --
Fast Boot to _boot.py only | 70 | 3.32 | 4.6
Send ESPNow Succeed | 119 | 7.68 | 10.7
Send ESPNow Failed | 145 | 15.4 | 21.4


</body>
</html>

The first column is milliseconds; is this the time for the re-transmissions?
If so how is it that in all measurements I've made so far the send return time has always been 25ms, ACK or NACK?


I am also seeing a greater battery drain then I expected. I won't be sure till I finish my testing that has resulted in more transmissions than would be normal. The units quiescent draw is 37ua.
 
Thanks glenn

**Reply by glenn20** [reply]
> Regarding https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowvsWifiEnergyUsage There's something puzzling about this table entry, specifically the "Send" times:
> 
> Test	Time (ms)	Charge (mC)	Energy (microWh)
> Fast Boot to _boot.py only	70	3.32	4.6
> Send ESPNow Succeed	119	7.68	10.7
> Send ESPNow Failed	145	15.4	21.4
> The first column is milliseconds; is this the time for the re-transmissions? If so how is it that in all measurements I've made so far the send return time has always been 25ms, ACK or NACK?

To understand that page you should look at https://github.com/glenn20/upy-esp32-experiments.

These are measurements of the time and energy usage to boot the ESP32 from deepsleep, send a message and go back to deepsleep. They assume all the boot optimisations from https://github.com/glenn20/upy-esp32-experiments/tree/main/OptimisingMicropythonBootTime and represent very best case scenarios.

If you are interested in saving battery power, there are some dramatic power savings to be made - on the ESP32 devices. I haven't done the same tests for esp8266. You also need a board that is designed for low power consumption in deepsleep (like most of the devices from UnexpectedMaker). Some boards have very poor power utilisation from other components on the board.

> I am also seeing a greater battery drain then I expected. I won't be sure till I finish my testing that has resulted in more transmissions than would be normal. The units quiescent draw is 37ua.

Is that the board consumption or the esp8266 consumption. If it is the board that is pretty damn good.

**Reply by mmliam** [reply]
> Is that the board consumption or the esp8266 consumption. If it is the board that is pretty damn good.

The 37ua figure is for ALL the circuitry in my remote unit; the sensor, external circuitry and the ESP-07 module. It's all do the external circuitry I designed that allows a sensor event detection to wake the ESP-07 from hibernation, then allows the ESP-07 to control it's own hibernation state once awaken by the sensor.  The unit (primarily the ESP-07) would be drawing around 3.5ma, if it were not in hibernation. Deepsleep doesn't work for the ESP-07 as there are other components within the module that are unaffected by the ESP8266 going into deepsleep; the entire module must be put in hibernation. The sensor alone draws 20ua; and that, of course, must stay active when the ESP-07 is in hibernation. BTW: The ESP32 is too large for the remote unit I designed; I use an ESP32 in my gateway device that doesn't have any size or power restrictions

> These are measurements of the time and energy usage to boot the ESP32 from deepsleep, send a message and go back to deepsleep. They assume all the boot optimisations from https://github.com/glenn20/upy-esp32-experiments/tree/main/OptimisingMicropythonBootTime and represent very best case scenarios.

When I make my timing measurements I start the timer just before the first send, and measure it after all sends have completed.
With the receiver off (no ACK's received) I measure 76ms for the 3 send attempts the unit makes. With the receiver on (ACK received) I measure 25ms (only one send).

Have you measured the time of the sends only?

 I interpreted the "Table" that says "Send ESPNow Succeed" & "Send ESPNow Fail" as measuring only the send time as I did. But even if the Table is including boot-time and other extraneous code, there is still a significant difference between success and failure; that's not what I'm seeing.

**Reply by glenn20** [reply]
>
> > These are measurements of the time and energy usage to boot the ESP32 from deepsleep, send a message and go back to deepsleep. They assume all the boot optimisations from https://github.com/glenn20/upy-esp32-experiments/tree/main/OptimisingMicropythonBootTime and represent very best case scenarios.
> 
> When I make my timing measurements I start the timer just before the first send, and measure it after all sends have completed.
> With the receiver off (no ACK's received) I measure 76ms for the 3 send attempts the unit makes. With the receiver on (ACK received) I measure 25ms (only one send).
> 
> Have you measured the time of the sends only?

Yes. As I have mentioned several times in the past, a successful send on esp32 takes about 2ms. A failed send takes 25ms. Which is completely consistent with expectations.


>  I interpreted the "Table" that says "Send ESPNow Succeed" & "Send ESPNow Fail" as measuring only the send time as I did. But even if the Table is including boot-time and other extraneous code, there is still a significant difference between success and failure; that's not what I'm seeing.

Again, as I have mentioned numerous times now, there was a busy loop in earlier versions of the code that meant it took at least 25ms to execute a send.

**Reply by mmliam** [reply]
> Yes. As I have mentioned several times in the past, a successful send on esp32 takes about 2ms. A failed send takes 25ms. Which is completely consistent with expectations.

That's interesting, I don't get any difference in the send return time (ACK, or NACK); it's always 25ms. This I consider necessary to maintain synchronization between all my remote units..

> Again, as I have mentioned numerous times now, there was a busy loop in earlier versions of the code that meant it took at least 25ms to execute a send.

Sorry, I missed you very last comment: So you're saying this "busy loop" was keeping the send return time consistent; ACK or NACK? Why was it changed (removing the much desired determinism it provided), and in what version of micropython did the change occur? 

BTW: Something I never mentioned before; when the ESP-07 makes a transmission the current draw spikes to about 300ma (actual transmission), drops to 70ma for a coded 5 sec delay, then goes into hibernation. This is consistent with the current draw reported by others with similar modules. I actually have a 470uF cap over the ESP-07 power, as recommended by some other users, to assist with the transmission current spikes.

**Reply by glenn20** [reply]
> So you're saying this "busy loop" was keeping the send return time consistent; ACK or NACK? Why was it changed (removing the much desired determinism it provided), and in what version of micropython did the change occur?

After sending a message with sync==True (the default) the code waits till the espressif software stack flags that an ACK has been received or has timed out (timeout is ~25ms - though not documented). After sending the message, the code would go into a loop, check if the ack has been received then invoke `mp_hal_delay_ms(BUSY_WAIT_MS)` before checking again. Unfortunately, I had forgotten that BUSY_WAIT_MS is set to 25ms (as this was the same delay in other busy loops to ensure other tasks get an opportunity to execute while busy-waiting). I have changed that code, so that it will return as soon as the ack is received.

I really think that the great majority of users would prefer that `send()` returns as soon as the ack is received.

**Reply by mmliam** [reply]
I understand the reason you made the change, but if it were me I would have included an extra (optional) parameter in the send(). That parameter being boolean defaulted to FALSE. For legacy code there would be no difference; a 25ms return for an ACK, or NACK. Set the optional parameter TRUE would result in an immediate return on ACK.

Alternatively, there could have been two versions of send(); not unlike recv() & irecv()

Best of both possible worlds :-)

**Reply by glenn20** [reply]
> I understand the reason you made the change, but if it were me I would have included an extra (optional) parameter in the send(). That parameter being boolean defaulted to FALSE. For legacy code there would be no difference; a 25ms return for an ACK, or NACK. Set the optional parameter TRUE would result in an immediate return on ACK.
> 
> Alternatively, there could have been two versions of send(); not unlike recv() & irecv()
> 
> Best of both possible worlds :-)

I understand why you would like that for your particular use case, but I reckon it would have zero chance of being accepted for merge in the PR (and I'd agree with that :-).

**Reply by mmliam** [reply]
> I understand why you would like that for your particular use case

I do suspect I'm not the only one; now I don't want to upgrade beyond 1.19.1 for concern it would mess-up the synchronism of sensors. I would need to create my own "busy loop" on send() True.

> I reckon it would have zero chance of being accepted for merge in the PR (and I'd agree with that :-).

Pity. I reckon if it were done in the beginning it would be painless; only an extra feature. As it's been said nothing ventured, nothing gained :-)


BTW: I've asked this earlier; in what version of micropython did you make the change to the send() return? 

Also, glenn please I hope you understand, I'm not trying to be overly critical, or difficult; I really do appreciate your opinion, expert help, and professionalism!
